feat: group creation api

dawnfire05 opened the pull request:
close #379 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- 그룹 인증을 위한 `GroupAuthBloc` 클래스 추가.
	- 그룹 생성 및 관리에 필요한 API 메서드 포함된 `GroupApi` 클래스 추가.
	- 그룹 및 역할 생성에 사용되는 데이터 모델 추가 (`CreateGroupModel`, `CreateRoleModel`, 등).
	- 그룹 생성 프로세스를 위한 BLoC 패턴 구현.
	- `AuthApi` 및 `BaseAuthApi`를 통한 인증 API 추가.
	- `AuthTokenModel` 데이터 모델 추가.
	- `ZiggleDio` 및 `GroupsDio`를 위한 새로운 HTTP 클라이언트 클래스 추가.
	- 그룹 생성 임시 데이터를 위한 `GroupCreateDraftEntity` 클래스 추가.
	- 사용자 인증을 위한 새로운 `GroupsRestAuthRepository` 및 `ZiggleRestAuthRepository` 클래스 추가.
	- 새로운 `GroupsCookieManager` 및 `ZiggleCookieManager` 클래스 추가하여 쿠키 관리 기능 향상.

- **Bug Fixes**
	- 그룹 생성 시 사용자 입력을 BLoC에 통합하여 상태 관리 개선.

- **Documentation**
	- 새로운 API 및 데이터 모델에 대한 문서화 추가.

- **Chores**
	- 코드 구조 개선 및 불필요한 의존성 제거.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dawnfire05 committed:
fix: group creation notion page button width

dawnfire05 committed:
feat: initialized group bloc

dawnfire05 committed:
feat: group api

dawnfire05 committed:
feat: add notion api

dawnfire05 committed:
feat: added part '*.g.dart'; to models

dawnfire05 committed:
feat: name add event

dawnfire05 committed:
feat: add all events of group creation

dawnfire05 committed:
feat: group create entity

dawnfire05 committed:
fix: notionPageId

dawnfire05 committed:
feat: implement group create bloc

dawnfire05 committed:
fix: group desription is required

dawnfire05 committed:
feat: implementation of GroupReposiitory

dawnfire05 committed:
feat: call createGroup function

dawnfire05 committed:
feat: group repository

dawnfire05 committed:
feat: description in group entity is not ?

dawnfire05 committed:
feat: annotate unused code

dawnfire05 committed:
feat: seperate dio

dawnfire05 commented:
@2paperstar 
dio 분리에 관한
a388227
커밋에 대해 리뷰 부탁드려요

dawnfire05 committed:
feat: add parameter in fcm api

dawnfire05 committed:
feat: add named parameter

dawnfire05 committed:
feat: dio module now lazySingleton

dawnfire05 committed:
fix: name of dio

2paperstar commented:
> @2paperstar dio 분리에 관한 [a388227](https://github.com/gsainfoteam/ziggle-flutter/commit/a3882275c17be644d0e84451271112206e40c66a) 커밋에 대해 리뷰 부탁드려요

구두로 설명한 내용을 조금 정리해보겠습니다
- injectable은 여러개의 module을 지원하지 않습니다. `di` 디렉토리 밑에 있는 모듈 파일을 대신 사용해야합니다
- typo가 있을 수 있기 때문에 `@Named` 대신에 Dio class를 extend해서 별도로 injection하기
	- 하지만 Dio가 abstract class여서 구현할 수 없기 때문에 똑같이 abstract class로 extend 한 후에 factory method 를 제공해주면 됩니다
- ApiChannel enum의 값은 staging, production 두개로 유지, 대신에 final String 필드를 하나 더 만들어서 ziggleBaseUrl, groupsBaseUrl을 별도로 받을 수 있게 구현
- Dio에는 AuthorizedInterceptor, CookieManager 둘다 넣어주기
- 각 Dio에 붙는 AuthorizeInterceptor의 코드는 같으나, TokenRepository의 코드가 달라져야 함.
    - factoryParam을 쓸 수도 있고, AuthorizedInterceptor를 abstact로 만든 뒤에 그것을 구현하는 class를 두개로 쪼개기

dawnfire05 committed:
feat: extend dioForNative

dawnfire05 committed:
feat: implement extended dio

dawnfire05 committed:
feat: divide api channel

dawnfire05 committed:
feat: implement api channel repository with divided api channel

dawnfire05 committed:
feat: update developer option bloc

dawnfire05 committed:
chore: remove unused import in module.dart

dawnfire05 committed:
feat: implement divided token repository

dawnfire05 committed:
feat: implement seperated authorize interceptor

dawnfire05 committed:
feat: use seperated interceptor

dawnfire05 committed:
fix: groups stage api url

dawnfire05 committed:
refactor: api channel in one file

dawnfire05 committed:
refactor : api channel repositoires

dawnfire05 committed:
refactor: remove unused imports

dawnfire05 committed:
refactor: simplified api channel

dawnfire05 committed:
feat: rest auth repository uses ziggleTokenRepository

dawnfire05 committed:
fix: channel.name

dawnfire05 committed:
feat: authorize_interceptor abstract class

dawnfire05 committed:
feat: group auth logic

dawnfire05 committed:
fix: constructor parameter to BaseAuthApi

dawnfire05 committed:
feat: rest auth repositories are singelton

dawnfire05 committed:
refactor: injection between auth ~~

2paperstar commented on lib/app/modules/groups/data/data_sources/remote/auth_api.dart:
```diff
@@ -17,7 +18,8 @@ abstract class AuthApi extends BaseAuthApi {
     @Query('code') String code, [
-    @Query('redirectUri') String redirectUri,
+    @Query('redirectUri')
+    String redirectUri = 'ziggle-idp-login-redirect://callback',
```
살짝 아쉬운거는 이게 strings.dart 파일에서 가져와졌으면 하는 것?

2paperstar commented on lib/app/modules/user/data/repositories/groups_rest_auth_repository.dart:
```diff
@@ -1,15 +1,18 @@
       AuthApi api,
-      GroupsFlutterSecureStorageTokenRepository tokenRepository,
+      @Named.from(GroupsFlutterSecureStorageTokenRepository)
+      FlutterSecureStorageTokenRepository tokenRepository,
```
super

2paperstar commented on lib/app/modules/user/data/repositories/rest_user_repository.dart:
```diff
@@ -18,7 +19,8 @@ class RestUserRepository implements UserRepository {
 
-  RestUserRepository(this._api, AuthRepository authRepository) {
+  RestUserRepository(this._api,
+      @Named.from(ZiggleRestAuthRepository) RestAuthRepository authRepository) {
```
auth repository를 field로 만들어주세요

2paperstar commented on lib/app/modules/user/data/repositories/ziggle_rest_auth_repository.dart:
```diff
@@ -1,15 +1,18 @@
   ZiggleRestAuthRepository(
       UserApi api,
-      ZiggleFlutterSecureStorageTokenRepository tokenRepository,
+      @Named.from(ZiggleFlutterSecureStorageTokenRepository)
+      FlutterSecureStorageTokenRepository tokenRepository,
```
super

dawnfire05 committed:
feat: idPRedirectUri is const

dawnfire05 committed:
fix: removed unused instance type

dawnfire05 committed:
refactor: created field

dawnfire05 committed:
feat: add logger in GroupsDio

dawnfire05 committed:
fix: modify models according to the API

dawnfire05 committed:
feat: notionPageId can be null

dawnfire05 committed:
feat: BlocProvider goes to the GroupCreationShell

dawnfire05 committed:
feat: nees Type

dawnfire05 committed:
refactor: uses arrow function

dawnfire05 committed:
feat: some properties can be null

dawnfire05 committed:
fix: locale setting

dawnfire05 committed:
fix/refactor: minor issues
- delete unused import
- Podfile.lock updated
- used @named for Web2OAuthRepository in RestAuthRepository

dawnfire05 committed:
feat: group creation logic tested

dawnfire05 commented on lib/app/modules/core/data/repositories/memory_api_channel_repository.dart:
```diff
@@ -14,12 +14,14 @@ import '../../domain/repositories/api_channel_repository.dart';
-    _localSubscription = _subject.listen(
-      (value) => _dio.options.baseUrl = value.baseUrl,
-    );
+  final ZiggleDio _ziggleDio;
+  final GroupsDio _groupsDio;
+
+  MemoryApiChannelRepository(this._ziggleDio, this._groupsDio) {
+    _localSubscription = _subject.listen((value) {
+      _ziggleDio.options.baseUrl = value.ziggleBaseUrl;
+      _groupsDio.options.baseUrl = value.groupsBaseUrl;
+    });
```
일리가 있네...
그러면 onRequest에서 그걸 설정하라는 말이야?
근데 사실 요청 중간에서 개발자 모드를 바꿀 일은 없잖어.
그런 경우는 상정하지 않아도 될 것 같은데?

dawnfire05 commented on lib/app/modules/user/data/repositories/ziggle_rest_auth_repository.dart:
```diff
@@ -0,0 +1,23 @@
+import 'package:ziggle/app/modules/user/data/repositories/ziggle_flutter_secure_storage_token_repository.dart';
+import 'package:ziggle/app/modules/user/data/repositories/ziggle_web_auth_2_oauth_repository.dart';
+
+@named
+@Singleton(as: RestAuthRepository)
+class ZiggleRestAuthRepository extends RestAuthRepository {
```
안돼. Named 파라미터를 사용하지 않으면 ziggle rest auth repository와 groups rest auth repository를 구분할 방법이 없어.
그렇지 않아?

dawnfire05 commented on lib/app/modules/user/presentation/bloc/group_auth_bloc.dart:
```diff
@@ -0,0 +1,33 @@
+
+part 'group_auth_bloc.freezed.dart';
+
+@injectable
+class GroupAuthBloc extends Bloc<GroupAuthEvent, GroupAuthState> {
+  final RestAuthRepository _repository;
+
+  GroupAuthBloc(@Named.from(GroupsRestAuthRepository) this._repository)
+      : super(const GroupAuthState.initial()) {
+    on<_Load>((event, emit) {});
+    on<_Login>(
+      (event, emit) async {
+        await _repository.login();
+      },
+    );
+  }
+}
```
오케이 그렇게 테스트 한 번 해볼게

dawnfire05 commented on lib/app/modules/user/data/repositories/groups_rest_auth_repository.dart:
```diff
@@ -0,0 +1,25 @@
+@named
+@Singleton(as: RestAuthRepository)
+class GroupsRestAuthRepository extends RestAuthRepository {
+  GroupsRestAuthRepository(
+      AuthApi api,
+      @Named.from(GroupsFlutterSecureStorageTokenRepository)
+      FlutterSecureStorageTokenRepository tokenRepository,
+      CookieManager cookieManager,
+      @Named.from(GroupsWebAuth2OauthRepository)
+      WebAuth2OAuthRepository oAuthRepository)
+      : super(
+            api: api,
+            tokenRepository: tokenRepository,
+            cookieManager: cookieManager,
+            oAuthRepository: oAuthRepository);
```
오케이 이것도 테스트 해볼게

dawnfire05 committed:
chore: solve conversations with coderabbit

dawnfire05 committed:
refactor: change solid class to abstract

dawnfire05 commented on lib/app/modules/groups/presentation/pages/group_creation_notion_page.dart:
```diff
@@ -59,7 +69,10 @@
         ),
         const SizedBox(height: 30),
-        ZiggleInput(hintText: context.t.group.creation.notion.hint),
+        ZiggleInput(
+          onChanged: (v) => setState(() => _notionPageId = v),
+          hintText: context.t.group.creation.notion.hint,
+        ),
```
그렇게 되면 쓸데없이 bloc에 요청을 너무 많이 보내지 않아?

dawnfire05 committed:
feat: resolve conversations

2paperstar commented:
@dawnfire05 pubspec.lock 파일 변경 되돌리고 다시 액션 돌려볼 수 있을까요?

dawnfire05 commented:
@2paperstar 넴

dawnfire05 committed:
revert pubspec.lock

2paperstar commented on ios/Podfile.lock:
얘도 다시 돌려줘요

2paperstar commented on lib/app/modules/core/data/dio/groups_dio.dart:
```diff
@@ -0,0 +1,20 @@
+@singleton
+class GroupsDio extends DioForNative {
+  GroupsDio(GroupsAuthorizeInterceptor authorizeInterceptor,
+      CookieManager cookieManager) {
```
쿠키 매니저도 같이 쓰면 안 되지 않아요?

2paperstar commented on lib/app/modules/core/data/dio/groups_dio.dart:
```diff
@@ -0,0 +1,20 @@
+      requestBody: true,
+      responseBody: true,
+      responseHeader: false,
+    ));
```
데브툴즈의 네트워크 탭 쓰세요

2paperstar commented on lib/app/modules/groups/data/data_sources/models/auth_token_model.dart:
여기 파일들 디렉토리가 잘못 되어 있어요

2paperstar commented on lib/app/modules/groups/data/data_sources/models/group_response_model.dart:
```diff
@@ -0,0 +1,23 @@
+
+@freezed
+class GroupResponseModel with _$GroupResponseModel {
+  factory GroupResponseModel(
```
const factory

2paperstar commented on lib/app/modules/groups/data/data_sources/models/group_response_model.dart:
```diff
@@ -0,0 +1,23 @@
+part 'group_response_model.g.dart';
+
+@freezed
+class GroupResponseModel with _$GroupResponseModel {
```
GroupModel 이랑 차이점이 뭐에요?

2paperstar commented on lib/app/modules/groups/data/repositories/rest_group_repository.dart:
```diff
@@ -0,0 +1,50 @@
+    File? image,
+  }) async {
+    if (name.trim().isEmpty) {
+      throw ArgumentError('그룹 이름은 비워둘 수 없습니다.');
```
에러는 i18n 을 고려해야 합니다
별도의 에러 클래스를 선언해서 그걸 던지고, 사용하는 쪽에서 받은 에러 타입에 따라 처리하도록 하면 깔끔합니다. 아니면 영어로 에러 메시지를 통일해주세요

2paperstar commented on lib/app/modules/groups/data/repositories/rest_group_repository.dart:
```diff
@@ -0,0 +1,50 @@
+      ));
+
+      if (image != null) {
+        // await _api.uploadImage(createdGroup.uuid, image);
```
TODO? WIP?

dawnfire05 commented:
@2paperstar 리뷰해주세용

2paperstar commented on lib/app/modules/groups/domain/entities/group_create_entity.dart:
```diff
@@ -0,0 +1,12 @@
+part 'group_create_entity.freezed.dart';
+
+@freezed
+class GroupCreateEntity with _$GroupCreateEntity {
```
draft entity라는게 이름에 보이면 좋겠어요

2paperstar commented on lib/app/modules/groups/presentation/blocs/group_bloc.dart:
```diff
@@ -0,0 +1,26 @@
+
+@freezed
+sealed class GroupEvent with _$GroupEvent {
+  const factory GroupEvent() = _New;
```
이거 잘못 정의된 것 같은데 확인해주세요

2paperstar commented on lib/app/modules/user/data/data_sources/remote/authorize_interceptor.dart:
```diff
@@ -1,33 +1,29 @@
-class AuthorizeInterceptor extends Interceptor {
-  final TokenRepository _repository;
+abstract class AuthorizeInterceptor extends Interceptor {
+  final FlutterSecureStorageTokenRepository repository;
```
이게 TokenRepository면 안 되나요?

2paperstar commented on lib/app/modules/user/presentation/bloc/developer_option_bloc.dart:
이 파일은 바꿀 필요가 없는 것 같은데, 맞죠?

2paperstar reviewed (changes requested):
인증 쪽에 print 함수가 남아있던데, 필요없으면 지워주세요

2paperstar reviewed (changes requested):
인증 쪽에 print 함수가 남아있던데, 필요없으면 지워주세요

dawnfire05 committed:
feat: developer option bloc no need to change

dawnfire05 committed:
refactor: deleted dio log

dawnfire05 committed:
feat: group response model and group model is same

dawnfire05 committed:
revert Podfile.lock

dawnfire05 committed:
fix: resolve conversations

dawnfire05 committed:
feat: move groups related auth logic

dawnfire05 committed:
refactor: remove unused import

dawnfire05 committed:
add cookie manager

dawnfire05 committed:
feat: add implemented cookie manager to dios

dawnfire05 committed:
feat: AuthorizeInterceptor onError

dawnfire05 committed:
fix: delete refresh mehtod

2paperstar commented on lib/app/modules/groups/data/repositories/rest_group_repository.dart:
```diff
@@ -0,0 +1,46 @@
+    File? image,
+  }) async {
+    if (name.trim().isEmpty) {
+      throw ArgumentError('Group name can' ' be empty');
```
?

2paperstar commented on lib/app/modules/groups/data/repositories/rest_group_repository.dart:
```diff
@@ -0,0 +1,46 @@
+
+      return createdGroup;
+    } catch (e) {
+      throw Exception(e);
```
딱히 하는 일이 없는데, try catch를 묶는 이유가 있나요?

2paperstar commented on lib/app/modules/user/data/data_sources/remote/authorize_interceptor.dart:
```diff
@@ -1,37 +1,13 @@
-  AuthorizeInterceptor(this._repository);
-
-  @override
-  void onError(DioException err, ErrorInterceptorHandler handler) async {
-    final dio = sl<Dio>();
-
-    final statusCode = err.response?.statusCode;
-    if (statusCode != 401) return handler.next(err);
-    final token = await _repository.token.first;
-    if (token == null) return handler.next(err);
-    if (err.requestOptions.retried) return handler.next(err);
-    err.requestOptions.retried = true;
-
-    try {
-      if (!(await _refresh())) return handler.next(err);
-      final retriedResponse = await dio.fetch(err.requestOptions);
-      return handler.resolve(retriedResponse);
-    } on DioException {
-      return super.onError(err, handler);
-    }
-  }
```
이거 구현을 냅두고 refreshToken? 같은 abstract 함수를 따로 둬요
이거 그대로 있으면 나중에 중복 될거라

2paperstar replied:
groups에서 refresh 구현 되기 전까지는 groups auth interceptor에서는 빈 함수로 두면 되겠죠?

2paperstar reviewed (changes requested):
조금만 더 봐주세요 ㅜㅜ 화이팅!!

dawnfire05 committed:
feat: resolve conversations on rest group repository

dawnfire05 committed:
feat: resolve conversations on auth inteptr

2paperstar commented on lib/app/modules/user/data/data_sources/remote/groups_authorize_interceptor.dart:
```diff
@@ -0,0 +1,26 @@
+  @override
+  Future<bool> refresh() {
+    // TODO: implement refresh
+    throw UnimplementedError();
```
일단은 토큰 지우는걸로 구현해두면 어때요?

2paperstar reviewed (dismissed):
코멘트 확인해주고 머지 합니다!

dawnfire05 committed:
feat: groups intercpt refresh function

dawnfire05 commented:
와 드디어
10.1 ~ 11.13
